tachyon/zk/plonk/vanishing/prover_vanishing_argument.h

```diff
@@ -0,0 +1,322 @@
+    // Sample a random polynomial of degree n - 1
+    const Evals random_eval = Evals::One();
+    const Poly random_poly = prover->domain()->IFFT(random_eval);
+
+    // Sample a random blinding factor
+    // TODO(TomTaehoonKim): Figure out why it is named |random_blind|.
+    F random_blind = F::Zero();
+
+    if (!prover->Commit(random_poly)) return false;
+
+    *out = Committed<PCSTy>{std::move(random_poly), random_blind};
+    return true;
+  }
+};
+
+template <typename PCSTy>
+class Committed {
+ public:
+  using F = typename PCSTy::Field;
+  using Poly = typename PCSTy::Poly;
+  using Domain = typename PCSTy::Domain;
+  using ExtendedDomain = typename PCSTy::ExtendedDomain;
+  using ExtendedPoly = typename PCSTy::ExtendedPoly;
+  using ExtendedEvals = typename PCSTy::ExtendedEvals;
+
+  constexpr Committed() = default;
+  constexpr Committed(Poly random_poly, const F& random_blind)
+      : random_poly_(std::move(random_poly)), random_blind_(random_blind) {}
+
+  constexpr Poly& random_poly() { return random_poly_; }
+  constexpr F& random_blind() const { return random_blind_; }
+
+  // This divides the polynomial (in the extended domain) by the vanishing
+  // polynomial of the 2ᵏ size domain.
+  ExtendedEvals DivideByVanishingPoly(const ExtendedEvals& evals,
+                                      const ExtendedDomain* extended_domain,
+                                      const Domain* domain) const {
+    CHECK_EQ(evals.NumElements(), extended_domain->size());
+
+    // Calculate ζ = g^((2ˢ * T) / 3)^2
+    // NOTE(TomTaehoonKim): Halo2 does not calculate ζ = g^((2ˢ * T) / 3), but
+    // the square of it.
+    CHECK_EQ(F::Config::kTrace % math::BigInt<F::kLimbNums>(3),
+             math::BigInt<F::kLimbNums>(0));
+    F zeta = F::FromMontgomery(F::Config::kSubgroupGenerator)
+                 .Pow(F(2).Pow(F::Config::kTwoAdicity).ToBigInt() *
+                      F::Config::kTrace / math::BigInt<F::kLimbNums>(3))
+                 .Square();
+
+    // Compute the evaluations of t(X) = Xⁿ - 1 in the coset evaluation domain.
+    // We don't have to compute all of them, because it will repeat.
+    std::vector<F> t_evaluations;
+    t_evaluations.reserve(size_t{1} << (extended_domain->log_size_of_group() -
+                                        domain->log_size_of_group()));
+    F orig = zeta.Pow(domain->size());
+    F step = extended_domain->group_gen().Pow(domain->size());
+    F cur = orig;
+    do {
+      t_evaluations.push_back(cur);
+      cur *= step;
+    } while (cur != orig);
+    CHECK_EQ(t_evaluations.size(),
+             size_t{1} << (extended_domain->log_size_of_group() -
+                           domain->log_size_of_group()));
+
+    // Subtract 1 from each to give us t_evaluations[i] = t(zeta *
+    // extended_omegaⁱ)
+    base::Parallelize(t_evaluations, [](absl::Span<F> chunk) {
+      for (F& coeff : chunk) {
+        coeff -= F::One();
+      }
+    });
```

Comment:
Could you move this to `UnivariateEvaluations`?

Reply:
It seems tightly coupled with translation, but I'm not sure we can relate a function and create one. I'll leave a comment.